service/history/query.go

```diff
@@ -0,0 +1,120 @@
+//
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package history
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/.gen/go/shared"
+)
+
+const (
+	queryStateBuffered queryState = iota
+	queryStateCompleted
+)
+
+type (
+	queryState int
+
+	query interface {
+		getQuerySnapshot() *querySnapshot
+		getQueryTermCh() <-chan struct{}
+		completeQuery(*shared.WorkflowQueryResult) error
+	}
+
+	queryImpl struct {
+		sync.RWMutex
+
+		id          string
+		queryInput  *shared.WorkflowQuery
+		queryResult *shared.WorkflowQueryResult
+		termCh      chan struct{}
+		state       queryState
+	}
+
+	querySnapshot struct {
+		id          string
+		queryInput  *shared.WorkflowQuery
+		queryResult *shared.WorkflowQueryResult
+		state       queryState
+	}
+)
+
+func newQuery(queryInput *shared.WorkflowQuery) query {
+	return &queryImpl{
+		id:          uuid.New(),
+		queryInput:  queryInput,
+		queryResult: nil,
+		termCh:      make(chan struct{}),
+		state:       queryStateBuffered,
+	}
+}
+
+func (q *queryImpl) getQuerySnapshot() *querySnapshot {
+	q.RLock()
+	defer q.RUnlock()
+
+	return &querySnapshot{
+		id:          q.id,
+		queryInput:  q.queryInput,
+		queryResult: q.queryResult,
+		state:       q.state,
+	}
+}
+
+func (q *queryImpl) getQueryTermCh() <-chan struct{} {
+	q.RLock()
+	defer q.RUnlock()
+
+	return q.termCh
+}
+
+func (q *queryImpl) completeQuery(queryResult *shared.WorkflowQueryResult) error {
+	q.Lock()
+	defer q.Unlock()
+
+	if q.state == queryStateCompleted {
+		return errors.New("query already completed")
+	}
+	if err := validateQueryResult(queryResult); err != nil {
+		return err
+	}
+	q.state = queryStateCompleted
+	q.queryResult = queryResult
+	close(q.termCh)
+	return nil
+}
+
+func validateQueryResult(queryResult *shared.WorkflowQueryResult) error {
+	if queryResult == nil {
+		return errors.New("query result is nil")
```

Comment:
internal server error?

Reply:
It does not really matter much because this metric is never returned to the user, it will only be emitted to log errors. But logically it is internal service error so I changed it to such.